File: app/src/main/java/me/echeung/moemoekyun/viewmodel/UserViewModel.java
package me.echeung.moemoekyun.viewmodel;

import androidx.databinding.Bindable;

import androidx.databinding.library.baseAdapters.BR;

import me.echeung.moemoekyun.client.model.User;

public class UserViewModel extends BaseViewModel {

    private User user;
    private String avatarUrl;
    private String bannerUrl;

    private boolean hasFavorites;

    public void reset() {
        setUser(null);
        setAvatarUrl(null);
        setBannerUrl(null);

        setRequestsRemaining(0);
        setHasFavorites(false);
    }

    @Bindable
    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
        setRequestsRemaining(user != null ? user.getRequestsRemaining() : 0);
        notifyPropertyChanged(BR.user);
    }

    @Bindable
    public String getAvatarUrl() {
        return avatarUrl;
    }

    public void setAvatarUrl(String avatarUrl) {
        this.avatarUrl = avatarUrl;
        notifyPropertyChanged(BR.avatarUrl);
    }

    @Bindable
    public String getBannerUrl() {
        return bannerUrl;
    }

    public void setBannerUrl(String bannerUrl) {
        this.bannerUrl = bannerUrl;
        notifyPropertyChanged(BR.bannerUrl);
    }

    @Bindable
    public int getRequestsRemaining() {
        if (user == null) {
            return 0;
        }

        return user.getRequestsRemaining();
    }

    public void setRequestsRemaining(int requestsRemaining) {
        if (user != null) {
            user.setRequestsRemaining(requestsRemaining);
        }
        notifyPropertyChanged(BR.requestsRemaining);
    }

    @Bindable
    public boolean getHasFavorites() {
        return hasFavorites;
    }

    public void setHasFavorites(boolean hasFavorites) {
        this.hasFavorites = hasFavorites;
        notifyPropertyChanged(BR.hasFavorites);
    }

}

File: app/src/main/kotlin/me/echeung/moemoekyun/viewmodel/UserViewModel.kt
package me.echeung.moemoekyun.viewmodel

import androidx.databinding.Bindable
import androidx.databinding.library.baseAdapters.BR
import me.echeung.moemoekyun.client.api.model.User

class UserViewModel : BaseViewModel() {

    @get:Bindable
    var user: User? = null
        set(user) {
            field = user
            notifyPropertyChanged(BR.user)
        }

    @get:Bindable
    var avatarUrl: String? = null
        set(avatarUrl) {
            field = avatarUrl
            notifyPropertyChanged(BR.avatarUrl)
        }

    @get:Bindable
    var bannerUrl: String? = null
        set(bannerUrl) {
            field = bannerUrl
            notifyPropertyChanged(BR.bannerUrl)
        }

    @get:Bindable
    var hasFavorites: Boolean = false
        set(hasFavorites) {
            field = hasFavorites
            notifyPropertyChanged(BR.hasFavorites)
        }

    fun reset() {
        user = null
        avatarUrl = null
        bannerUrl = null

        hasFavorites = false
    }
}
